src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -560,6 +564,156 @@ private async ValueTask SendRstStreamAsync(int streamId, Http2ProtocolErrorCode
+                }
+            }
+        }
+
+        private void WriteHeaders(HttpRequestMessage request)
+        {
+            Debug.Assert(_headerBuffer.ActiveMemory.Length == 0);
+
+            // HTTP2 does not support Transfer-Encoding: chunked, so disable this on the request.
+            request.Headers.TransferEncodingChunked = false;
+
+            HttpMethod normalizedMethod = HttpMethod.Normalize(request.Method);
+
+            // TODO: ISSUE 31307: Use static table for pseudo-headers
+            WriteHeader(":method", normalizedMethod.Method);
+            WriteHeader(":scheme", "https");
+
+            string authority;
+            if (request.HasHeaders && request.Headers.Host != null)
+            {
+                authority = request.Headers.Host;
+            }
+            else
+            {
+                authority = request.RequestUri.IdnHost;
+                if (!request.RequestUri.IsDefaultPort)
+                {
+                    // TODO: Avoid allocation here by caching this on the connection or pool
+                    authority += ":" + request.RequestUri.Port;
+                }
+            }
+
+            WriteHeader(":authority", authority);
+            WriteHeader(":path", request.RequestUri.GetComponents(UriComponents.PathAndQuery | UriComponents.Fragment, UriFormat.UriEscaped));
+
+            if (request.HasHeaders)
+            {
+                WriteHeaderCollection(request.Headers);
+            }
+
+            // Determine cookies to send.
+            if (_pool.Settings._useCookies)
+            {
+                string cookiesFromContainer = _pool.Settings._cookieContainer.GetCookieHeader(request.RequestUri);
+                if (cookiesFromContainer != string.Empty)
+                {
+                    WriteHeader(HttpKnownHeaderNames.Cookie, cookiesFromContainer);
+                }
+            }
+
+            if (request.Content == null)
+            {
+                // Write out Content-Length: 0 header to indicate no body,
+                // unless this is a method that never has a body.
+                if (normalizedMethod.MustHaveRequestBody)
+                {
+                    // TODO: ISSUE 31307: Use static table for Content-Length
+                    WriteHeader("Content-Length", "0");
+                }
+            }
+            else
+            {
+                WriteHeaderCollection(request.Content.Headers);
+            }
+        }
+
+        private async ValueTask SendHeadersAsync(int streamId, HttpRequestMessage request)
+        {
+            // Note, HEADERS and CONTINUATION frames must be together, so hold the writer lock across sending all of them.
+            // We also serialize usage of the header encoder and the header buffer this way.
+            await _writerLock.WaitAsync().ConfigureAwait(false);
+
+            // Generate the entire header block, without framing, into the connection header buffer.
+            WriteHeaders(request);
```

Comment:
If this method throws we'll still be holding the `_writerLock`. Presumably that means we can no longer use the connection at all, even if we just run in to issues with the headers of a single request. Is that intentional?

Reply:
Good catch, not intentional. I'll move the try block up to cover this as well. I don't think WriteHeaders should really fail (aside from OOM), but there's no reason not to just protect against it.